DEV-5708 and DEV-5709: add mvpp2 timestamping

paul-sirin opened the pull request:
DEV-5709: add support for HW RX TS in mvpp2
Adds:
 * enables HW RX TS using offload flags, but configuration is done
   separately in Linux using Timestamping API
 * detection of mvpp2 TAI using Linux ethtool
 * PHC read using Linux PTP subsystem
 * handling of timestamps with convertion from u32 to u64 nsec
 * alarm worker that triggers PHC gettimex64() in Linux every 2 seconds

DEV-5708: add support for HW TX TS in mvpp2
Adds:
 * includes unification changes for HW RX TS (see prev. commit)
 * enables HW TX TS using offload flags on per-packet basis, but configuration
   is done separately in Linux using Timestamping API before VPP start
 * adds data structures to keep track of egress queue of timestamps
 * adds 1ms poll for new timestamps, masks IRQ for new timestamps (see TODOs)
 * adds hardcoded TX descriptor setup (see TODOs)
 * adds support for 1 TX timesramp queue (see TODOs)

TODO:
 * add support for PTP packet classifier and parser
 * add support for PTP IRQs
 * add support for 2 TX timesramp queues

Commit:
DEV-5709: add support for HW RX TS in mvpp2

Adds:
 * enables HW RX TS using offload flags, but configuration is done
   separately in Linux using Timestamping API
 * detection of mvpp2 TAI using Linux ethtool
 * PHC read using Linux PTP subsystem
 * handling of timestamps with convertion from u32 to u64 nsec
 * alarm worker that triggers PHC gettimex64() in Linux every 2 seconds

shmuelhazan commented on drivers/net/mvpp2/mrvl_ptp.c:
```diff
@@ -0,0 +1,346 @@
+	/* Iterate over 0..31 */
+	queue->next = (i + 1) & 31;
+	rte_pktmbuf_free(queue->mbuf[i]);
+	queue->mbuf[i] = rte_pktmbuf_clone(mbuf, mbuf->pool);
```
why clone the buffer? (instead of adding a refcount)

paul-sirin replied:
It's a proper way to increment refcount in DPDK

shmuelhazan replied:
You are right. The naming is just strange.

shmuelhazan commented on drivers/net/mvpp2/mrvl_ptp.c:
```diff
@@ -0,0 +1,346 @@
+		}
+	}
+
+	if (rte_eal_alarm_set(MS_PER_S, mvpp2_isr_handle_ptp_queue,
```
where does this alarm runs on? another thread? if so, we probably need to consider synchronization here.

paul-sirin replied:
all alarms run on DPDK interrupt thread. it seems we need sync for shared hwtstamp_queue's mbuf refcount, because I'm not sure if it is thread safe

shmuelhazan replied:
I think that the refcount is probably ok. I suspect more about other shared data (e.g. the tx TS queue)

paul-sirin replied:
I still doubt that we need to keep mbufs: Linux does that to use skbuf->sk socket to push back timestamps there (back to app), but I don't see such possibility for mbuf. My current idea is to simplify that and remove mbuf references list. What do you think?

paul-sirin replied:
added spinlock for queue->next

shmuelhazan commented on drivers/net/mvpp2/mrvl_ptp.c:
```diff
@@ -0,0 +1,346 @@
+	queue = &priv->tx_hwtstamp_queue[nq];
+
+	while (1) {
+		u32 ts = pp2_ppio_get_outq_ts(priv->ppio, nq, &it);
```
I could not find source for this function; in any case, it is unclear how the HW knows to associate "it" with a packet; as mvpp2_tx_hw_tstamp does not forward "i" to the HW.

paul-sirin replied:
I've moved this to MUSDK as a patch because it needs access to library internals. In mvpp2_tx_hwtstamp "i" is encoded in ptp descriptor (and set to "tx_desc") and then should emerge on ptp register "r0" as "it"
see https://github.com/siklu/devices/pull/1358/files#diff-f1403f7ce1669f51f8d75231f0e513f5860db0b6e6ff02d0fd16124d024760fd

shmuelhazan commented on drivers/net/mvpp2/mrvl_ethdev.c:
```diff
@@ -2875,6 +2907,7 @@ mrvl_tx_pkt_burst(void *txq, struct rte_mbuf **tx_pkts, uint16_t nb_pkts)
 		 * in case unsupported ol_flags were passed
 		 * do not update descriptor offload information
 		 */
+		mvpp2_tx_hw_tstamp(q->priv, &descs[i], mbuf);
```
I think that this function should be inlined as it is in the fast path (in the future, we probably need to consider inlining MUSDK calls too).
Also, we should think how to minimize performance impact on non-ptp usages.

paul-sirin replied:
done

paul-sirin replied:
For RX TS per-port config ol_flag is used to enable, for TX TS per-packet ol_flag is used to perform stamping

shmuelhazan commented on drivers/net/mvpp2/mrvl_ethdev.c:
```diff
@@ -2654,6 +2672,20 @@ mrvl_rx_pkt_burst(void *rxq, struct rte_mbuf **rx_pkts, uint16_t nb_pkts)
 				mrvl_desc_to_ol_flags(&descs[i],
 						      mbuf->packet_type);
 
+		if (mrvl_enable_ts[mbuf->port] && (q->priv->tai != CLOCK_INVALID)) {
```
maybe unlikely would be good here.

paul-sirin replied:
I've hidden `mrvl_enable_ts[]` into ptp build unit and exposed this check as function and wrapped in `unlikely`

shmuelhazan replied:
that does not help here. The compiler can't see that unlikely.

shmuelhazan commented on drivers/net/mvpp2/mrvl_ethdev.c:
```diff
@@ -2654,6 +2672,20 @@ mrvl_rx_pkt_burst(void *rxq, struct rte_mbuf **rx_pkts, uint16_t nb_pkts)
 						      mbuf->packet_type);
 
+		if (mrvl_enable_ts[mbuf->port] && (q->priv->tai != CLOCK_INVALID)) {
+			u32 timestamp =
+				rte_le_to_cpu_32(pp2_ppio_inq_desc_get_timestamp(&descs[i]));
+			if (mvpp22_tai_tstamp(q->priv->tai, timestamp,
+					mrvl_timestamp_dynfield(mbuf))) {
+				MRVL_LOG(ERR, "failed to convert pkt timestamp:0x%x", timestamp);
+			} else {
+				mbuf->ol_flags |= mrvl_timestamp_rx_dynflag;
+				MRVL_LOG(DEBUG,
+						"Port %d RX pkt timestamp %u convert to %"PRIu64 " nsec\n",
+						q->port_id, timestamp, *mrvl_timestamp_dynfield(mbuf));
+			}
```
maybe need to move that to another (inline) function

Commit:
DEV-5708: add support for HW TX TS in mvpp2

Adds:
 * includes unification changes for HW RX TS (see prev. commit)
 * enables HW TX TS using offload flags on per-packet basis, but configuration
   is done separately in Linux using Timestamping API before VPP start
 * adds data structures to keep track of egress queue of timestamps
 * adds 1ms poll for new timestamps, masks IRQ for new timestamps (see TODOs)
 * adds hardcoded TX descriptor setup (see TODOs)
 * adds support for 1 TX timesramp queue (see TODOs)

TODO:
 * add support for PTP packet classifier and parser
 * add support for PTP IRQs
 * add support for 2 TX timesramp queues

shmuelhazan commented on drivers/net/mvpp2/mrvl_ptp.c:
```diff
@@ -0,0 +1,431 @@
+	return true;
+}
+
+inline bool mvpp2_is_rx_ts_enabled(struct mrvl_rxq *q)
```
why is it inline if it is only used out of the file? Maybe need to move the definition to the header file.

paul-sirin replied:
I don't want to expose global static variables, so just put unlikely outside and make it non-inline

shmuelhazan commented on drivers/net/mvpp2/mrvl_ptp.c:
```diff
@@ -0,0 +1,431 @@
+
+bool mvpp2_enable_rx_ts(uint16_t port_id)
+{
+	if (rte_mbuf_dyn_rx_timestamp_register(&mrvl_timestamp_dynfield_offset,
+				&mrvl_timestamp_rx_dynflag)) {
+		MRVL_LOG(ERR, "Failed to register Rx timestamp field/flag");
+		return false;
+	}
```
are you sure that we want to do that for each port?

paul-sirin replied:
It just sets up proper mask, see further it is assigned to mbuf->ol_flag when RX TS decoded and set to mbuf_dynfield

shmuelhazan replied:
Not sure I understand you. Wy we are calling rte_mbuf_dyn_rx_timestamp_register for each port?

paul-sirin replied:
added check for non-negative offset before registering dynfield

shmuelhazan replied:
How does it solve the issue? if the offset is positive or zero it will still re-register it. if not it will print an error

paul-sirin replied:
taking in account: The logical disjunction operator || tests if at least one of two expressions it true. If the first expression is true, then the second expression is not evaluated. (see https://www.gnu.org/software/gnu-c-manual/gnu-c-manual.html#Logical-Operators )
we have following cases:
if ((offset=-1 < 0) || (register()=-1 != 0)) print_err
if ((offset=-1 < 0) || (register()=0 != 0))
if ((offset=0..N < 0) || /* no eval */)

shmuelhazan replied:
just this discussion means that probably it is better to rewrite it in a more clear way. 
In any case, I don't think it is true. 
```
	if (mrvl_timestamp_dynfield_offset < 0 ||
	    rte_mbuf_dyn_rx_timestamp_register(&mrvl_timestamp_dynfield_offset,
					       &mrvl_timestamp_rx_dynflag)) {
		MRVL_LOG(ERR, "Failed to register Rx timestamp field/flag");
		return false;
	}
```
Will run ` rte_mbuf_dyn_rx_timestamp_register(&mrvl_timestamp_dynfield_offset, &mrvl_timestamp_rx_dynflag)` if mrvl_timestamp_dynfield_offset >= 0.

paul-sirin replied:
I see the issue, that should be &&

paul-sirin replied:
then here are options:
```
	if (mrvl_timestamp_dynfield_offset < 0 &&
	    rte_mbuf_dyn_rx_timestamp_register(&mrvl_timestamp_dynfield_offset,
					       &mrvl_timestamp_rx_dynflag)) {
		MRVL_LOG(ERR, "Failed to register Rx timestamp field/flag");
		return false;
	}
```
and
```
	if (mrvl_timestamp_dynfield_offset < 0) {
	    if (rte_mbuf_dyn_rx_timestamp_register(&mrvl_timestamp_dynfield_offset,
					          &mrvl_timestamp_rx_dynflag)) {
		MRVL_LOG(ERR, "Failed to register Rx timestamp field/flag");
		return false;
            }
	}
```

shmuelhazan replied:
let's do the second -- I think it is more clear.

shmuelhazan commented on drivers/net/mvpp2/mrvl_ptp.c:
```diff
@@ -0,0 +1,431 @@
+		close(fd);
+		return CLOCK_INVALID;
+	}
+	if (clock_adjtime(clkid, &tx)) {
+		close(fd);
+		return CLOCK_INVALID;
+	}
```
unclear why this is needed.

paul-sirin replied:
I assume yes, because this code was inspired by kernel

shmuelhazan replied:
where in the kernel? adjtime uses here garbage values.

paul-sirin replied:
excuse me, not by kernel, but by linuxptp. see https://github.com/richardcochran/linuxptp/blob/master/phc.c#L42

paul-sirin replied:
If I understand correctly it just tries to apply adjastement to the clock with a step of 0

shmuelhazan replied:
oh - I missed the memset.

shmuelhazan commented on drivers/net/mvpp2/mrvl_ptp.c:
```diff
@@ -0,0 +1,431 @@
+	int ret = rte_eal_alarm_set(2*US_PER_S, mvpp2_phc_get_current_time,
+			(void *)priv);
+	if (ret == 0) {
+		priv->scheduled_get_time_alarm;
```
something is missing here?

paul-sirin replied:
fixed

shmuelhazan commented on drivers/net/mvpp2/mrvl_ptp.c:
```diff
@@ -0,0 +1,431 @@
+		return false;
+	}
+
+#if 0
+	unsigned int mtype;
+	struct ptp_header *hdr;
+	// TODO: add classifier support
+	type = ptp_classify_raw(mbuf);
+	if (!type)
+		return false;
+
+	hdr = ptp_parse_header(mbuf, type);
+	if (!hdr)
+		return false;
+
+	skb_shinfo(mbuf)->tx_flags |= SKBTX_IN_PROGRESS;
+#endif
```
maybe need to remove it.

paul-sirin replied:
I just wanted to keep it as reference of functionality that is missing at current development stage

shmuelhazan commented on drivers/net/mvpp2/mrvl_ptp.c:
```diff
@@ -0,0 +1,431 @@
+	*ptp_desc &= rte_cpu_to_le_32(~MVPP22_PTP_DESC_MASK_LOW);
+}
+
+inline bool mvpp2_tx_hw_tstamp(
```
here too: either move it to header or make it non-inline.

paul-sirin replied:
non-inline, but I'd better move ol_flag check outside

shmuelhazan replied:
maybe it should be inline. these functions are being called frequently.

Commit:
DEV-5708: PR fixes

Commit:
DEV-5708: don't register ts dynfield multiple times

shmuelhazan commented on drivers/net/mvpp2/mrvl_ptp.c:
```diff
@@ -0,0 +1,433 @@
+static uint64_t mrvl_timestamp_rx_dynflag;
+static int mrvl_timestamp_dynfield_offset = -1;
+
+static inline rte_mbuf_timestamp_t *
```
you probably want to move it (and many other functions - if not all) to the header.

paul-sirin replied:
I didn't want, but it seems I will have to

Commit:
DEV-5708: inline functions in fast-path

Commit:
DEV-5708: fix debug logs
